src/re/multi_grid/matern.py

```diff
@@ -0,0 +1,321 @@
+        5.97489612400613639965e2,
+        2.11688757100572135698e2,
+        2.52070205858023719784e1,
+    ]
+)
+_QQ1 = jnp.array(
+    [
+        1.0,
+        7.42373277035675149943e1,
+        1.05644886038262816351e3,
+        4.98641058337653607651e3,
+        9.56231892404756170795e3,
+        7.99704160447350683650e3,
+        2.82619278517639096600e3,
+        3.36093607810698293419e2,
+    ]
+)
+
+_Z1 = 1.46819706421238932572e1
+_Z2 = 4.92184563216946036703e1
+_THPIO4 = 2.35619449019234492885  # 3*pi/4
+_SQ2OPI = 0.79788456080286535588  # sqrt(2/pi)
+
+
+def _j1_small(x):
+    z = x * x
+    w = jnp.polyval(_RP1, z) / jnp.polyval(_RQ1, z)
+    w = w * x * (z - _Z1) * (z - _Z2)
+    return w
+
+
+def _j1_large_c(x):
+    w = 5.0 / x
+    z = w * w
+    p = jnp.polyval(_PP1, z) / jnp.polyval(_PQ1, z)
+    q = jnp.polyval(_QP1, z) / jnp.polyval(_QQ1, z)
+    xn = x - _THPIO4
+    p = p * jnp.cos(xn) - w * q * jnp.sin(xn)
+    return p * _SQ2OPI / jnp.sqrt(x)
+
+
+def j1(x):
+    """
+    Bessel function of order one - using the implementation from CEPHES,
+    translated to Jax.
+
+    Pure JAX implementation for Bessel function from
+    https://github.com/benjaminpope/sibylla/blob/main/notebooks/bessel_test.ipynb .
+    """
+    return jnp.sign(x) * jnp.where(
+        jnp.abs(x) < 5.0, _j1_small(jnp.abs(x)), _j1_large_c(jnp.abs(x))
+    )
+
+
+FourierIntegralGrid = namedtuple(
+    "FourierIntegralGrid",
+    (
+        "mode_lengths",
+        "mode_binbounds",
+        "min_dist",
+        "max_dist",
+        "num",
+        "ndim",
+        "weights",
+    ),
+)
+
+
+def make_integral_grid(
+    min_dist: float, max_dist: float, num: int, ndim: int, normalize: bool = True
+) -> FourierIntegralGrid:
+    mode_lengths = np.geomspace(1.0 / max_dist, 1.0 / min_dist, num, endpoint=False)
+    mode_lengths = np.insert(mode_lengths, 0, 0.0)
+    # Binbounds
+    lk = np.log(mode_lengths[1:])
+    dlk = (np.log(max_dist) - np.log(min_dist)) / num
+    lk = np.append(lk - 0.5 * dlk, lk[-1] + 0.5 * dlk)
+    mode_binbounds = np.insert(np.exp(lk), 0, 0.0)
+    # norm_weights
+    wgt = None
+    if normalize:
+        if ndim == 1:
+            fkr = sici(mode_binbounds * max_dist)[0]
+        elif ndim == 2:
+            fkr = 1.0 - j0(mode_binbounds * max_dist)
+        elif ndim == 3:
+            fkr = sici(mode_binbounds * max_dist)[0] - np.sin(mode_binbounds * max_dist)
+        else:
+            raise NotImplementedError
+        wgt = fkr[1:] - fkr[:-1]
+        if (ndim == 1) or (ndim == 3):
+            wgt *= 2.0 / np.pi
+    return FourierIntegralGrid(
+        min_dist=min_dist,
+        max_dist=max_dist,
+        num=num,
+        ndim=ndim,
+        mode_lengths=mode_lengths,
+        mode_binbounds=mode_binbounds,
+        weights=wgt,
+    )
+
+
+def spectrum2covariance(fig: FourierIntegralGrid, spec: npt.NDArray, *, normalize=True):
+    fct = [np.pi, 2.0 * np.pi, 2.0 * np.pi**2]
+    assert spec.size == fig.num + 1
```

Comment:
With a non-static fig (as it is the default below) I believe these asserts will cause issues while tracing